File: 4 semestre/Paradigmas de Linguagem de Programação/10_outubro/aula 23-10-25/java x c++ x c#/SistemaBiblioteca.java
/*
Paradigmas de Linguagem de Programação
Aula: 23-10-25
Slide 9: Diferença entre Java vs C++ vs C#
Exemplo em Java - Sistema de Biblioteca
*/

import java.util.*;

class Livro {
    private String titulo;
    private String autor;
    private boolean emprestado;

    public Livro(String titulo, String autor) {
        this.titulo = titulo;
        this.autor = autor;
        this.emprestado = false;
    }

    public String getTitulo() {
        return titulo;
    }

    public String getAutor() {
        return autor;
    }

    public boolean isEmprestado() {
        return emprestado;
    }

    public void emprestar() {
        emprestado = true;
    }

    public void devolver() {
        emprestado = false;
    }

    public String toString() {
        return titulo + " - " + autor + (emprestado ? " (Emprestado)" : " (Disponível)");
    }
}

class Biblioteca {
    private List<Livro> livros;

    public Biblioteca() {
        livros = new ArrayList<>();
    }

    public void adicionarLivro(String titulo, String autor) {
        livros.add(new Livro(titulo, autor));
    }

    public void listarLivros() {
        for (int i = 0; i < livros.size(); i++) {
            System.out.println((i + 1) + ". " + livros.get(i));
        }
    }

    public void buscarPorTitulo(String titulo) {
        for (Livro livro : livros) {
            if (livro.getTitulo().equalsIgnoreCase(titulo)) {
                System.out.println("Encontrado: " + livro);
                return;
            }
        }
        System.out.println("Livro não encontrado.");
    }

    public void emprestarLivro(int indice) {
        if (indice >= 0 && indice < livros.size()) {
            Livro livro = livros.get(indice);
            if (!livro.isEmprestado()) {
                livro.emprestar();
                System.out.println("Livro emprestado com sucesso.");
            } 
            else {
                System.out.println("Livro já está emprestado.");
            }
        } 
        else {
            System.out.println("Índice inválido.");
        }
    }

    public void devolverLivro(int indice) {
        if (indice >= 0 && indice < livros.size()) {
            Livro livro = livros.get(indice);
            if (livro.isEmprestado()) {
                livro.devolver();
                System.out.println("Livro devolvido com sucesso.");
            } 
            else {
                System.out.println("Livro não está emprestado.");
            }
        } 
        else {
            System.out.println("Índice inválido.");
        }
    }
}

public class SistemaBiblioteca {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Biblioteca biblioteca = new Biblioteca();
        int opcao;

        do {
            System.out.println("\n--- Menu ---");
            System.out.println("1. Adicionar livro");
            System.out.println("2. Listar livros");
            System.out.println("3. Buscar por título");
            System.out.println("4. Emprestar livro");
            System.out.println("5. Devolver livro");
            System.out.println("0. Sair");
            System.out.print("Escolha uma opção: ");
            opcao = scanner.nextInt();
            scanner.nextLine();

            switch (opcao) {
                case 1:
                    System.out.print("Título: ");
                    String titulo = scanner.nextLine();
                    System.out.print("Autor: ");
                    String autor = scanner.nextLine();
                    biblioteca.adicionarLivro(titulo, autor);
                    break;
                case 2:
                    biblioteca.listarLivros();
                    break;
                case 3:
                    System.out.print("Digite o título: ");
                    String busca = scanner.nextLine();
                    biblioteca.buscarPorTitulo(busca);
                    break;
                case 4:
                    biblioteca.listarLivros();
                    System.out.print("Digite o número do livro para emprestar: ");
                    int emprestar = scanner.nextInt() - 1;
                    biblioteca.emprestarLivro(emprestar);
                    break;
                case 5:
                    biblioteca.listarLivros();
                    System.out.print("Digite o número do livro para devolver: ");
                    int devolver = scanner.nextInt() - 1;
                    biblioteca.devolverLivro(devolver);
                    break;
                case 0:
                    System.out.println("Encerrando...");
                    break;
                default:
                    System.out.println("Opção inválida.");
            }
        } while (opcao != 0);

        scanner.close();
    }
}

File: 4 semestre/Paradigmas de Linguagem de Programação/10_outubro/aula 23-10-25/java x c++ x c#/sistema_biblioteca.cc
/*
Paradigmas de Linguagem de Programação
Aula: 23-10-25
Slide 9: Diferença entre Java vs C++ vs C#
Exemplo em C++ - Sistema de Biblioteca
*/

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Livro {
private:
    string titulo;
    string autor;
    bool emprestado;

public:
    Livro(string t, string a) : titulo(t), autor(a), emprestado(false) {}

    string getTitulo() const { return titulo; }
    string getAutor() const { return autor; }
    bool isEmprestado() const { return emprestado; }

    void emprestar() { emprestado = true; }
    void devolver() { emprestado = false; }

    void exibir() const {
        cout << titulo << " - " << autor << (emprestado ? " (Emprestado)" : " (Disponível)") << endl;
    }
};

class Biblioteca {
private:
    vector<Livro> livros;

public:
    void adicionarLivro(string titulo, string autor) {
        livros.push_back(Livro(titulo, autor));
    }

    void listarLivros() const {
        for (size_t i = 0; i < livros.size(); ++i) {
            cout << i + 1 << ". ";
            livros[i].exibir();
        }
    }

    void buscarPorTitulo(string titulo) const {
        for (const auto& livro : livros) {
            if (livro.getTitulo() == titulo) {
                cout << "Encontrado: ";
                livro.exibir();
                return;
            }
        }
        cout << "Livro não encontrado." << endl;
    }

    void emprestarLivro(int indice) {
        if (indice >= 0 && indice < livros.size()) {
            if (!livros[indice].isEmprestado()) {
                livros[indice].emprestar();
                cout << "Livro emprestado com sucesso." << endl;
            } 
            else {
                cout << "Livro já está emprestado." << endl;
            }
        } 
        else {
            cout << "Índice inválido." << endl;
        }
    }

    void devolverLivro(int indice) {
        if (indice >= 0 && indice < livros.size()) {
            if (livros[indice].isEmprestado()) {
                livros[indice].devolver();
                cout << "Livro devolvido com sucesso." << endl;
            } else {
                cout << "Livro não está emprestado." << endl;
            }
        } else {
            cout << "Índice inválido." << endl;
        }
    }
};

int main() {
    Biblioteca biblioteca;
    int opcao;

    do {
        cout << "\n--- Menu ---\n";
        cout << "1. Adicionar livro\n";
        cout << "2. Listar livros\n";
        cout << "3. Buscar por título\n";
        cout << "4. Emprestar livro\n";
        cout << "5. Devolver livro\n";
        cout << "0. Sair\n";
        cout << "Escolha uma opção: ";
        cin >> opcao;
        cin.ignore();

        switch (opcao) {
            case 1: {
                string titulo, autor;
                cout << "Título: ";
                getline(cin, titulo);
                cout << "Autor: ";
                getline(cin, autor);
                biblioteca.adicionarLivro(titulo, autor);
                break;
            }
            case 2:
                biblioteca.listarLivros();
                break;
            case 3: {
                string busca;
                cout << "Digite o título: ";
                getline(cin, busca);
                biblioteca.buscarPorTitulo(busca);
                break;
            }
            case 4: {
                biblioteca.listarLivros();
                int indice;
                cout << "Digite o número do livro para emprestar: ";
                cin >> indice;
                biblioteca.emprestarLivro(indice - 1);
                break;
            }
            case 5: {
                biblioteca.listarLivros();
                int indice;
                cout << "Digite o número do livro para devolver: ";
                cin >> indice;
                biblioteca.devolverLivro(indice - 1);
                break;
            }
            case 0:
                cout << "Encerrando..." << endl;
                break;
            default:
                cout << "Opção inválida." << endl;
        }
    } while (opcao != 0);

    return 0;
}
